gpii/node_modules/flowManager/src/PCPChannel.js

```diff
@@ -0,0 +1,275 @@
+"use strict";
+
+
+var fluid = require("infusion");
+var gpii = fluid.registerNamespace("gpii");
+
+fluid.registerNamespace("gpii.flowManager");
+
+// A mixin grade applied to the lifecycleManager's session by the pcpChannel
+fluid.defaults("gpii.pcpChannel.sessionBinder", {
+    modelRelay: {
+        pcpChannel: {
+            source: { // Not "" because of FLUID-6192
+                segs: []
+            },
+            target: "{flowManager}.pcpChannel.model",
+            singleTransform: {
+                type: "gpii.pcpChannel.sessionToPCP",
+                ontologyMetadata: "{ontologyHandler}.ontologyMetadata"
+            },
+            // compensate for FLUID-6194
+            backward: "never",
+            forward: {
+                // avoid notifying the client for the init of LifecycleManagerSession, simplifies test and clients
+                excludeSource: "init"
+            }
+        }
+    },
+    modelListeners: {
+        updatePreferences: {
+            path: "{flowManager}.pcpChannel.model.settingControls",
+            includeSource: "PCP",
+            funcName: "gpii.pcpChannel.updatePreferences",
+            args: ["{flowManager}", "{pcpChannel}", "{flowManager}.lifecycleManager", "{change}.value"]
+        },
+        updateContext: {
+            path: "{flowManager}.pcpChannel.model.activeContextName",
+            includeSource: "PCP",
+            funcName: "gpii.pcpChannel.updateContext",
+            args: ["{flowManager}.contextManager", "{pcpChannel}", "{change}.value"]
+        }
+    }
+});
+
+fluid.defaults("gpii.pcpChannel", {
+    gradeNames: ["fluid.modelComponent"],
+    members: {
+        outputBlocked: null
+    },
+    distributeOptions: {
+        // Not {lifecycleManager session} because of FLUID-6193
+        target: "{flowManager lifecycleManager session}.options.gradeNames",
+        record: "gpii.pcpChannel.sessionBinder"
+    },
+    listeners: {
+        "{lifecycleManager}.events.onSessionStop": {
+            namespace: "pcpChannel",
+            funcName: "gpii.pcpChannel.sessionStop",
+            args: "{pcpChannel}"
+        }
+    }
+});
+
+gpii.pcpChannel.updatePreferences = function (flowManager, pcpChannel, lifecycleManager, newPrefs) {
+    fluid.log("Received preferences update from PCP UI of ", newPrefs);
+
+    // Grab current session
+    var session = lifecycleManager.getSession(lifecycleManager.getActiveSessionTokens());
+
+    // tweak payload to be in NP set format and update session
+    var contextualPrefs = {};
+    fluid.each(newPrefs, function (spec, pref) {
+        fluid.set(contextualPrefs, fluid.pathUtil.parseEL(pref), spec.value);
+    });
+
+    var prefcpy = fluid.copy(pcpChannel.model.preferences);
+    fluid.set(prefcpy, ["contexts", session.model.activeContextName, "preferences"], contextualPrefs);
```

Comment:
Is `prefcpy` actually needed? It seems that it is created, the preferences are added to it and subsequently it is not used.

Reply:
ah - good point, forgot to delete that after a little refactoring